internal/app/secretless/handlers/mysql/backend.go

```diff
@@ -70,273 +118,310 @@ func (h *Handler) ConfigureBackend() (err error) {
-	serverDoesntSupportSSL := serverHandshake.ServerCapabilities & protocol.ClientSSL == 0
-	if err != nil {
+func (mc *ConnectionPhase) validateServerSSL() {
+	if mc.err != nil {
 		return
 	}
 
-	// STEP: LoginRequest. Client => Secretless
-	//////////////////////////////////////////////
+	if mc.clientRequestedSSL() && !mc.serverSupportsSSL() {
+		mc.err = ErrNoTLS
+	}
+}
 
-	// Intercept response from client
-	clientHandshakeResponsePacket, err := h.ClientRead()
-	if err != nil {
-		return err
+func (mc *ConnectionPhase) readClientHandshakeResponse() {
+	if mc.err != nil {
+		return
 	}
 
-	if requestedSSL && serverDoesntSupportSSL {
-		return &protocol.Error{
-			Code:       protocol.CRSSLConnectionError,
-			SQLSTATE:   protocol.ErrorCodeInternalError,
-			Message:    ErrNoTLS.Error(),
-		}
+	rawResponse := mc.readClientPacket()
+	if mc.err != nil {
+		return
 	}
 
-	// Parse intercepted client response
 	// TODO: client requesting SSL results ERROR 2026 (HY000): SSL connection error: protocol version mismatch
-
-	clientHandshakeResponse, err := protocol.UnpackHandshakeResponse41(clientHandshakeResponsePacket)
-	if err != nil {
-		return err
+	mc.clientHandshakeResponse, mc.err = protocol.UnpackHandshakeResponse41(rawResponse)
+}
+func (mc *ConnectionPhase) overrideClientCapabilities() {
+	if mc.err != nil {
+		return
 	}
 
+	// TODO: after kumbi's done, change below to method calls
+
 	// TODO: add tests cases for authentication plugins support
 	// Disable CapabilityFlag for authentication plugins support
-	clientHandshakeResponse.CapabilityFlags &^= protocol.ClientPluginAuth
+	mc.clientHandshakeResponse.CapabilityFlags &^= protocol.ClientPluginAuth
 
 	// TODO: add tests cases for client secure connection
 	// Enable CapabilityFlag for client secure connection
-	clientHandshakeResponse.CapabilityFlags |= protocol.ClientSecureConnection
+	mc.clientHandshakeResponse.CapabilityFlags |= protocol.ClientSecureConnection
 
 	// Ensure CapabilityFlag is set when using TLS
-	if requestedSSL {
-		clientHandshakeResponse.CapabilityFlags |= protocol.ClientSSL
+	if mc.clientRequestedSSL() {
+		mc.clientHandshakeResponse.CapabilityFlags |= protocol.ClientSSL
 	}
 
-	// Inject credentials into client response
-	if err = protocol.InjectCredentials(clientHandshakeResponse, serverHandshake.Salt, h.BackendConfig.Username, h.BackendConfig.Password); err != nil {
+}
+
+func (mc *ConnectionPhase) injectCredentials() {
+	if mc.err != nil {
 		return
 	}
 
-	// This format of this packet is described here:
+	// TODO: change this to method call on clientHandshakeResponse when Kumbi's work done
+	mc.err = protocol.InjectCredentials(
+		mc.clientHandshakeResponse,
+		mc.serverHandshake.Salt,
+		mc.connectionDetails.Username,
+		mc.connectionDetails.Password,
+	)
+}
+
+// TODO: All of this should be moved to kumbi's code
+func (mc *ConnectionPhase) maybeUpgradeToSSL() {
+	if mc.err != nil {
+		return
+	}
+
+	if !mc.clientRequestedSSL() {
```

Comment:
why not just call the method `upgradeToSSL` and move this check to `ConnectionPhase.Run()`? I don't love the use of `maybe` in a method name

Reply:
`Run()` is only concerned with the higher level logic, very much by design.  I'm open to other names or strategies but I don't want to mar `Run()` with an an if statement relevant only to one method.  I considered `OptionallyUpgradeToSSL` or `HandleClientSSLRequests`, etc -- do you like those better?  I do not like `UpgradeToSSLIfRequested`, though...

Reply:
I think I prefer `HandleClientSSLRequests`